providers/class-two-factor-backup-codes.php

```diff
@@ -177,15 +177,23 @@ public function user_options( $user ) {
 							// Update counter.
 							$( '.two-factor-backup-codes-count' ).html( response.data.i18n.count );
 
-							// Build the download link.
-							var txt_data = 'data:application/text;charset=utf-8,' + '\n';
-							txt_data += response.data.i18n.title.replace( /%s/g, document.domain ) + '\n\n';
+							// Build the downloaded file contents.
+							var txt_data = response.data.i18n.title.replace( /%s/g, document.domain ) + '\n\n';
 
 							for ( i = 0; i < response.data.codes.length; i++ ) {
 								txt_data += i + 1 + '. ' + response.data.codes[ i ] + '\n';
 							}
 
-							$( '#two-factor-backup-codes-download-link' ).attr( 'href', encodeURI( txt_data ) );
+							txt_data = encodeURIComponent( txt_data );
+							// encodeURIComponent() Does not encode -_.!~*'()
+							txt_data = txt_data.replace(/[!'()~*]/g, function(c) {
```

Comment:
The comment mentions `-_.` in addition to `!'()~*` included in the current regex -- should those be included too?

Reply:
Good catch!

`encodeURI()` will not encode `~!@#$&*()=:/,;?+'`.
`encodeURUComponent()` will not encode `~!*()'` (ie. everything above that's not here are encoded)
The subset that's missing are `-_.` which neither encode, but are safe left unencoded.

```suggestion
							// encodeURIComponent() does not encode -_.!~*'(), encode the problematic characters
							txt_data = txt_data.replace(/[!'()~*]/g, function(c) {
```